ng-dev/misc/validate-licenses/validate.ts

```diff
@@ -0,0 +1,115 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import { init, ModuleInfo, ModuleInfos } from 'license-checker';
+import spdx from 'spdx-satisfies';
+
+
+// A general note on some disallowed licenses:
+// - CC0
+//    This is not a valid license. It does not grant copyright of the code/asset, and does not
+//    resolve patents or other licensed work. The different claims also have no standing in court
+//    and do not provide protection to or from Google and/or third parties.
+//    We cannot use nor contribute to CC0 licenses.
+// - Public Domain
+//    Same as CC0, it is not a valid license.
+
+
+/** List of established allowed licenses for depdenencies. */
+const allowedLicenses = [
+  // Notice licenses
+  'MIT',
+  'ISC',
+  'Apache-2.0',
+  'Python-2.0',
+  'Artistic-2.0',
+  'BSD-2-Clause',
+  'BSD-3-Clause',
+  'BSD-4-Clause',
+  'Zlib',
+  'AFL-2.1',
+  'CC-BY-3.0',
+  'CC-BY-4.0',
+
+  // Unencumbered
+  'Unlicense',
+  'CC0-1.0',
+  '0BSD',
+];
+
+/** Known name variations of SPDX licenses. */
+const licenseReplacements = new Map<string, string>([
+  // Just a longer string that our script catches. SPDX official name is the shorter one.
+  ['Apache License, Version 2.0', 'Apache-2.0'],
+  ['Apache2', 'Apache-2.0'],
+  ['Apache 2.0', 'Apache-2.0'],
+  ['Apache v2', 'Apache-2.0'],
+
+  // Alternate syntax
+  ['AFLv2.1', 'AFL-2.1'],
+
+  // BSD is BSD-2-clause by default.
+  ['BSD', 'BSD-2-Clause'],
+]);
+
+interface ExpandedModuleInfo extends ModuleInfo {
+  name: string;
+  allowed: boolean;
+}
+
+export interface LicenseCheckResult { valid: boolean, packages: ExpandedModuleInfo[], maxPkgNameLength: number };
+
+
+export async function checkAllLicenses(start: string): Promise<LicenseCheckResult> {
+  return new Promise((resolve, reject) => {
+    let maxPkgNameLength = 0;
+    init({ start }, (err: Error, pkgInfoObject: ModuleInfos) => {
```

Comment:
It seems like this tool relies on the local `node_modules` to be installed, compared to e.g. analyzing the lock file and consulting the registry (which I would find nicer -- but also this runs rarely I think).

What would happen if the local node modules are not installed? should we have a good check for that?

Reply:
Realistically if node_modules is not installed neither is `ng-dev`, but I am unsure if we should just rely on that fact.